src/search/pdbs/pattern_collection_generator_multiple_cegar.cc

```diff
@@ -0,0 +1,300 @@
+#include "pattern_collection_generator_multiple_cegar.h"
+
+#include "cegar.h"
+#include "pattern_database.h"
+#include "utils.h"
+
+#include "../option_parser.h"
+#include "../plugin.h"
+
+#include "../utils/countdown_timer.h"
+#include "../utils/markup.h"
+#include "../utils/rng.h"
+#include "../utils/rng_options.h"
+
+#include <vector>
+
+using namespace std;
+
+namespace pdbs {
+PatternCollectionGeneratorMultipleCegar::PatternCollectionGeneratorMultipleCegar(
+    options::Options &opts)
+    : max_pdb_size(opts.get<int>("max_pdb_size")),
+      max_collection_size(opts.get<int>("max_collection_size")),
+      wildcard_plans(opts.get<bool>("wildcard_plans")),
+      cegar_max_time(opts.get<double>("max_time")),
+      verbosity(opts.get<utils::Verbosity>("verbosity")),
+      rng(utils::parse_rng_from_options(opts)),
+      random_seed(opts.get<int>("random_seed")),
+      stagnation_limit(opts.get<double>("stagnation_limit")),
+      blacklist_trigger_percentage(opts.get<double>("blacklist_trigger_percentage")),
+      blacklist_on_stagnation(opts.get<bool>("blacklist_on_stagnation")),
+      total_max_time(opts.get<double>("total_max_time")) {
+}
+
+PatternCollectionInformation PatternCollectionGeneratorMultipleCegar::generate(
+    const shared_ptr<AbstractTask> &task) {
+    if (verbosity >= utils::Verbosity::NORMAL) {
+        utils::g_log << "Generating patterns using the multiple CEGAR algorithm" << endl;
+    }
+
+    TaskProxy task_proxy(*task);
+    utils::CountdownTimer timer(total_max_time);
+
+    // Store the set of goals in random order.
+    vector<FactPair> goals = get_goals_in_random_order(task_proxy, rng);
+
+    // Store the non-goal variables for potential blacklisting.
+    int num_vars = task_proxy.get_variables().size();
+    vector<int> non_goal_variables;
+    non_goal_variables.reserve(num_vars - goals.size());
+    for (int var_id = 0; var_id < num_vars; ++var_id) {
+        bool is_goal_var = false;
+        for (const FactPair &goal : goals) {
+            if (var_id == goal.var) {
+                is_goal_var = true;
+                break;
+            }
+        }
+        if (!is_goal_var) {
+            non_goal_variables.push_back(var_id);
+        }
+    }
+
+    if (verbosity >= utils::Verbosity::DEBUG) {
+        utils::g_log << "goal variables: ";
+        for (auto goal : goals) {
```

Comment:
I try to avoid using auto unless the class type is clear and the code gets much more readable with auto (e.g., in the case of iterators). In a case like this one, I'd go with "FactProxy".

Reply:
It is `FactPair` but yes, you are absolutely right. This is a remainder of what Alex wrote.